### ISSUE TITLE
Update testing framework

### ISSUE THREAD

**Issue by nwu63** [mdolab/pyoptsparse#145]
## Type of issue
What types of issue is it?
_Select the appropriate type(s) that describe this issue_

- Testing

## Description
We should update the overall testing framework:
- evaluate if we want to add other optimization problems
- share code among tests -- most of them are doing similar things and we should reduce code duplication
- improve overall code coverage

**Comment by matbesancon** [comment]
Not sure if this is related, but testing seems to block when running the command, and seems not to block but with no test being run without failing with the verbose flag, complete error:
```
./test/test_snopt_bugfix.py:TestSNOPTBug.test_opt_bug1 ... None (00:00:0.00, 0 MB)
./test/test_hs015.py:TestHS15.test_ipopt ... None (00:00:0.00, 0 MB)
./test/test_hs015.py:TestHS15.test_nlpqlp ... None (00:00:0.00, 0 MB)
./test/test_hs015.py:TestHS15.test_paropt ... None (00:00:0.00, 0 MB)
./test/test_hs015.py:TestHS15.test_psqp ... None (00:00:0.00, 0 MB)
./test/test_hs015.py:TestHS15.test_slsqp ... None (00:00:0.00, 0 MB)
./test/test_hs015.py:TestHS15.test_snopt ... None (00:00:0.00, 0 MB)
./test/test_snopt_bugfix.py:TestSNOPTBug.test_opt_bug_print_2con ... None (00:00:0.00, 0 MB)
./test/test_hs015.py:TestHS15.test_conmin ... None (00:00:0.00, 0 MB)
./test/test_snopt_bugfix.py:TestSNOPTBug.test_opt ... None (00:00:0.00, 0 MB)
./test/test_require_mpi_env_var.py:TestRequireMPIEnvVar.test_do_not_use_mpi ... FAIL (00:00:0.00, 31 MB)
Traceback (most recent call last):
  File "/home/mbesancon/.local/lib/python3.6/site-packages/testflo/test.py", line 425, in _try_call
    func()
  File "./test/test_require_mpi_env_var.py", line 42, in test_do_not_use_mpi
    import pyoptsparse.pyOpt_MPI
ModuleNotFoundError: No module named 'pyoptsparse'

./test/test_require_mpi_env_var.py:TestRequireMPIEnvVar.test_no_mpi_requirement_given ... FAIL (00:00:0.00, 31 MB)
Traceback (most recent call last):
  File "/home/mbesancon/.local/lib/python3.6/site-packages/testflo/test.py", line 425, in _try_call
    func()
  File "./test/test_require_mpi_env_var.py", line 32, in test_no_mpi_requirement_given
    import pyoptsparse.pyOpt_MPI
ModuleNotFoundError: No module named 'pyoptsparse'

./test/test_require_mpi_env_var.py:TestRequireMPIEnvVarOnParOpt.test_no_mpi_requirement_given_check_paropt ... SKIP (00:00:0.00, 31 MB)
('Optimizer not available:', 'paropt')
./test/test_require_mpi_env_var.py:TestRequireMPIEnvVarOnParOpt.test_require_mpi_check_paropt ... SKIP (00:00:0.00, 31 MB)
('Optimizer not available:', 'paropt')
./test/test_require_mpi_env_var.py:TestRequireMPIEnvVar.test_require_mpi ... SKIP (00:00:0.00, 31 MB)
mpi4py not available, skipping test.
./test/test_tp109.py:TestTP109.test_autorefine ... None (00:00:0.00, 0 MB)
./test/test_tp109.py:TestTP109.test_slsqp ... None (00:00:0.00, 0 MB)
./test/test_tp109.py:TestTP109.test_snopt ... None (00:00:0.00, 0 MB)
./test/test_require_mpi_env_var.py:TestRequireMPIEnvVarOnParOpt.test_do_not_use_mpi_check_paropt ... SKIP (00:00:0.00, 31 MB)
('Optimizer not available:', 'paropt')
./test/test_optProb.py:TestOptProb.test_mappings ... None (00:00:0.00, 0 MB)
./test/test_optProb.py:TestOptProb.test_setDV_VarGroup ... None (00:00:0.00, 0 MB)
./test/test_optProb.py:TestOptProb.test_setDV_getDV ... None (00:00:0.00, 0 MB)
./test/test_hs071.py:TestHS71.test_conmin ... None (00:00:0.00, 0 MB)
./test/test_hs071.py:TestHS71.test_ipopt ... None (00:00:0.00, 0 MB)
./test/test_hs071.py:TestHS71.test_nlpqlp ... None (00:00:0.00, 0 MB)
./test/test_hs071.py:TestHS71.test_paropt ... None (00:00:0.00, 0 MB)
./test/test_hs071.py:TestHS71.test_psqp ... None (00:00:0.00, 0 MB)
./test/test_hs071.py:TestHS71.test_slsqp ... None (00:00:0.00, 0 MB)
./test/test_hs071.py:TestHS71.test_slsqp_scaling_offset_optProb ... None (00:00:0.00, 0 MB)
./test/test_hs071.py:TestHS71.test_slsqp_setDV ... None (00:00:0.00, 0 MB)
./test/test_hs071.py:TestHS71.test_snopt_setDVFromHist ... None (00:00:0.00, 0 MB)
./test/test_hs071.py:TestHS71.test_snopt ... None (00:00:0.00, 0 MB)
./test/test_snopt_user_termination.py:TestUserTerminationStatus.test_sens ... None (00:00:0.00, 0 MB)
./test/test_snopt_user_termination.py:TestUserTerminationStatus.test_obj ... None (00:00:0.00, 0 MB)
./test/test_nsga2_multi_objective.py:TestNSGA2.test_opt ... None (00:00:0.00, 0 MB)
./test/test_large_sparse.py:TestLargeSparse.test_snopt ... None (00:00:0.00, 0 MB)
/home/mbesancon/Documents/crptests/pyoptsparse/pyoptsparse/pyOpt_MPI.py:63: UserWarning: mpi4py could not be imported. mpi4py is required to use the parallel gradient analysis and parallel objective analysis for non-gradient based optimizers. Continuing using a dummy MPI module from pyOptSparse.
  warnings.warn(warn)


The following tests were skipped:
test_require_mpi_env_var.py:TestRequireMPIEnvVar.test_require_mpi
test_require_mpi_env_var.py:TestRequireMPIEnvVarOnParOpt.test_do_not_use_mpi_check_paropt
test_require_mpi_env_var.py:TestRequireMPIEnvVarOnParOpt.test_no_mpi_requirement_given_check_paropt
test_require_mpi_env_var.py:TestRequireMPIEnvVarOnParOpt.test_require_mpi_check_paropt


The following tests failed:
test_require_mpi_env_var.py:TestRequireMPIEnvVar.test_do_not_use_mpi
test_require_mpi_env_var.py:TestRequireMPIEnvVar.test_no_mpi_requirement_given


Passed:  0
Failed:  2
Skipped: 4


Ran 36 tests using 4 processes
Wall clock time:   00:00:0.55
```

**Comment by matbesancon** [comment]
Linking the review issue:
https://github.com/mdolab/pyoptsparse/issues/145

**Comment by nwu63** [comment]
Hi Mathieu, would you mind opening a new issue for this so that we may discuss there? This issue is more about the overall scope of testing, and refactoring some existing tests.

**Comment by matbesancon** [comment]
Deleted comment that was my mistake

**Comment by nwu63** [comment]
Closed partially by #251, and superseded by #256.
